Local Token Decoding

chmoder opened the pull request:
Hi, we found this library when integrating keycloak and django.  It is a simple, easy to use library, thanks for sharing it!  One thing we noticed is that there are three HTTP requests from the middleware to keycloak.

Two to `/token/introspect`
One to `/openid-connect/userinfo`

The total time for all three requests is around **.80** seconds.

- adds django caching
- adds setting to use local token decoding instead of keycloak endpoints (default to use http interface)
- adds conditonal checks in `introspect`, `is_token_active`, `roles_from_token`,  and `userinfo` to use local decoding or http interface
- add jwks endpoint and method (with caching)
- adds token decode method

Commit:
add cryptography as a dependency

Commit:
add setting to use introspection or not

Commit:
add imports

Commit:
add use introspection to KeycloakConnect

Commit:
add use introspection to the middleware

Commit:
add keycloak certs endpoint

Commit:
add jwks endpoint method

Commit:
add jwks endpoint method

Commit:
use decode or introspect to determine if token is active

Commit:
add token decode method

Commit:
add token decode method

Commit:
use introspection or token decode to get user info

Commit:
fix requires list

Commit:
pass options to decoder

Commit:
add audience and options to decode parameters

Commit:
rename certs to jwks

Commit:
add django caching

Commit:
use django caching for jwks call

Commit:
lower default cache timeout

Commit:
update readme

Commit:
fix setting cache key name

Commit:
add dependency

Commit:
set default value for use_introspection

Commit:
add local decode to userinfo

Commit:
pass raise exception to decode

Commit:
bubble up error if token is not able to be decoded

Commit:
impl non-api vue handling

Commit:
test case for jwks with no exception

marcelo225 commented:
@chmoder, nice contributing!
I'll check this out soon.
Thanks for helping us!

marcelo225 commented:
@chmoder, could you change at [setup.py file](https://github.com/InsuranceToolkits/django-keycloak-auth/blob/master/setup.py#L11) the new version 0.9.9?

marcelo225 commented on setup.py:
```diff
@@ -14,6 +14,8 @@
     # Project uses reStructuredText, so ensure that the docutils get
     # installed or upgraded on the target machine
     install_requires=[
+        "cryptography",
```
@chmoder, Could you freeze these newly added libraries? I am afraid there might be some problems when someone uses this library. What do you think?

chmoder replied:
Great suggestions!  I pinned the version of pyJWT to the latest or anything that comes after.  And pyJWT has required cryptography>=3.4.0 for ~2 years now so I pinned ours to that dependency as well.

marcelo225 commented on setup.py:
```diff
@@ -14,6 +14,8 @@
     # installed or upgraded on the target machine
     install_requires=[
+        "cryptography",
+        "PyJWT",
```
This could be changed on this line as well

marcelo225 commented:
@chmoder, I've updated a #28 PR into the master branch recently. Could you merge it into your branch?

chmoder committed:
Merge branch 'marcelo225:master' into master

Commit:
bump version

Commit:
pin versions of cryptography and pyJWT

Commit:
clarify local_decode option in readme

Commit:
make jwks endpoint cache ttl a setting

chmoder commented:
> @chmoder, I've updated a #28 PR into the master branch recently. Could you merge it into your branch?

@marcelo225, I merged django-keycloak-auth master branch into this fork.  All your suggestions should be contributed as well.

I [added caching](https://github.com/InsuranceToolkits/django-keycloak-auth/blob/master/django_keycloak_auth/settings.py#L170) for the `jwks` endpoint.  Today I made the TTL for that a setting and updated documentation.  

Let me know what you think, the idea comes from [min-time-between-jwks-requests](https://www.keycloak.org/docs/latest/securing_apps/) and related documentation.  It may even be helpful for the `well_known_endpoint`.

Commit:
wording update

marcelo225 commented:
@chmoder , The new lib version 0.9.9 is now available. 
Thanks for contributing to it!
